commons/com.b2international.commons/src/com/b2international/commons/io/PathUtils.java

```diff
@@ -0,0 +1,271 @@
+		checkNotNull(path);
+		final String fileName = path.getFileName().toString();
+		final int dotIndex = fileName.lastIndexOf('.');
+		return dotIndex == -1 ? "" : fileName.substring(dotIndex + 1);
+	}
+
+	/**
+	 * Get the name of the file without the extension
+	 */
+	public static String getFileNameWithoutExtension(Path path) {
+		checkNotNull(path);
+		final String fileName = path.getFileName().toString();
+		final int dotIndex = fileName.lastIndexOf('.');
+		return dotIndex == -1 ? fileName : fileName.substring(0, dotIndex);
+	}
+
+	/**
+	 * Find a single file in either the root of a folder or the root of a zip archive using the filter expression
+	 */
+	public static Optional<Path> findFile(Path source, String pathMatcherExpression) throws IOException {
+		return findFiles(source, pathMatcherExpression, false).stream().findFirst();
+	}
+
+	/**
+	 * Find all files in either the root of a folder or the root of a zip archive using the filter expression
+	 */
+	public static Collection<Path> findFiles(Path source, String pathMatcherExpression) throws IOException {
+		return findFiles(source, pathMatcherExpression, false);
+	}
+
+	/**
+	 * Find a single file recursively in either a folder or a zip archive using the filter expression
+	 */
+	public static Optional<Path> findFileRecursive(Path source, String pathMatcherExpression) throws IOException {
+		return findFiles(source, pathMatcherExpression, true).stream().findFirst();
+	}
+
+	/**
+	 * Find all files recursively in either a folder or a zip archive using the filter expression
+	 */
+	public static Collection<Path> findFilesRecursive(Path source, String pathMatcherExpression) throws IOException {
+		return findFiles(source, pathMatcherExpression, true);
+	}
+
+	private static Collection<Path> findFiles(Path source, String pathMatcherExpression, boolean recursive) throws IOException {
+
+		if (Files.isDirectory(source)) {
+			return findFiles(source.getFileSystem(), Set.of(source), pathMatcherExpression, recursive);
+		} else if (getFileExtension(source).equals("zip")) {
+			try (FileSystem zipFileSystem = FileSystems.newFileSystem(source)) {
+				return findFiles(zipFileSystem, zipFileSystem.getRootDirectories(), pathMatcherExpression, recursive);
+			}
+		}
+
+		throw new UnsupportedOperationException("Cannot find file matches if source is not a directory or an archive");
+
+	}
+
+	private static Collection<Path> findFiles(FileSystem fileSystem, Iterable<Path> directoriesToIterate, String pathMatcherExpression, boolean recursive) throws IOException {
+
+		final Set<Path> matches = newHashSet();
+
+		final PathMatcher pathMatcher = fileSystem.getPathMatcher("glob:**/" + pathMatcherExpression);
+
+		for (final Path directory : directoriesToIterate) {
+			if (recursive) {
+				Files.walk(directory).filter(path -> pathMatcher.matches(path)).forEach(matches::add);
+			} else {
+				Files.list(directory).filter(path -> pathMatcher.matches(path)).forEach(matches::add);
```

Comment:
Since both `Files.walk(directory)` and `Files.list(directory)` return `Stream<Path>`, the following common portion of each line could be moved after the `if` block:

```java
final Stream<Path> paths;

if (recursive) { 
  paths = Files.walk(directory);
} else { 
  paths = Files.list(directory);
}

paths.filter(pathMatcher::matches).forEach(matches::add);
```